### ISSUE TITLE
Some thoughts on being nicer to beginners

### ISSUE THREAD

**Issue by eevee** [sirjuddington/SLADE#361]
I'm writing some stuff about getting into Doom mapping, for people who've never done it before and know absolutely nothing.  Jotting down possible beginner stumbling blocks as I come across them:

* First-time setup asks for base resource archives.  There are a couple obvious places to try to auto-detect them from, like the Steam and GOG default installation directories.  Could also try to look for an existing ZDoom or GZDoom config file and get the iwad paths from there.

* The last stage of first-time setup asks for a nodebuilder.  A beginner has no idea what a nodebuilder is, where to get one, or which one to use for which engine (can you even use ZDBSP nodes with vanilla Doom?).  If they're mapping for ZDoom, they won't even _need_ a nodebuilder, since ZDoom has one built in — but SLADE will still pop up a cryptic warning when they try to save their map.

* The default map settings are Chex Quest for the game, "(none)" for the port, and no base resource archive.  Pretty much anything else would be an improvement.  The game, for example, could be detected based on the known base resources.  ZDoom has even [done most of the work for us](https://github.com/rheit/zdoom/blob/master/wadsrc/static/iwadinfo.txt).

**Comment by Gaerzi** [comment]
Autodetecting IWADs is a good idea and I'm working on it. I'll get at least the content of [DOOMWADDIR & DOOMWADPATH](http://doomwiki.org/wiki/Environment_variables) plus the Steam and GOG.com install folders. I may also [borrow from Eternity](https://github.com/team-eternity/eternity/blob/master/source/d_findiwads.cpp) to find from various default pre-digital distribution directories.

For node builders, if DB2 is detected (its registry key was created) then we can fill it automatically from <db dir>\Compilers\Nodebuilders; it's distributed with bsp-w32.exe, deepbsp.exe, glbsp.exe, zdbsp.exe, and ZenNode.exe. But of course, like Doom Builder itself, that'll be only for Windows systems.

Detecting the game configuration from the base resource archive is possible, but I wonder how clever we should try to be about it. Should we change the config if the BRA is changed? The BRA is the config is changed?

**Comment by eevee** [comment]
I was delighted to discover a couple days ago that ZDoom can auto-detect a Steam-installed Doom II on OS X and Linux, too.

Can SLADE just ship with ACC (+ `zcommon`) and ZDBSP as well?  I'm pretty sure SLADE is putting out releases more frequently than those projects, so it's not in any danger of lagging behind, and an expert user can just point it at a more recent release if absolutely necessary.

It seems to me that the BRA and game config are nearly inextricably linked, and maybe we should just choose/filter the BRA automatically based on the game.  Is there any reason you would want them to differ?  I suppose you might want to make an IWAD completely from scratch, but that's a special case of having _no_ BRA.

Also I realize that if we ever do actually read from `zdoom.pk3` (#314), that's effectively a second BRA tied to the _port_.  If BRA handling changes, that might be worth considering too.

**Comment by Gaerzi** [comment]
Well for a start, "Doom 2" alone has three different base resource archives (Doom2.wad, TNT.wad, Plutonia.wad), and then we can add two more from Freedoom (FreeDM.wad, Freedoom2.wad). That's the biggie here, a same configuration can have multiple valid BRAs.

**Comment by eevee** [comment]
There's a tension between "don't be smarter than the user" and "guide the user away from obvious mistakes", I know.  Absolute worst case, though, you could still choose "none" as the BRA and open whatever you like as a regular archive.  Offhand I don't think the BRA is special to the map editor, and the "game" isn't used anywhere else, right?

**Comment by Gaerzi** [comment]
The game is used at the moment only for the map editor, but there are other features which possibly could use it later. An example: click on a THINGS lump and it'll display a table including type and flags, as raw number values; an enhancement of this feature could display "Imp; Hard Skill, Ambush" instead of "3001; 12". Likewise, Maintenance->Replace in Maps could provide a drop down of thing types. I'd also like to write in a feature to generate thing tables for Doom Wiki articles ([example](http://doomwiki.org/wiki/E1M1:_Hangar_%28Doom%29#Things)) and it'd be more handy if it doesn't need entering the map editor for that.

The BRA is essential for the texture editor and for the picture viewer.

**Comment by Gaerzi** [comment]
So, parsing the Steam config.vdf isn't going to be as easy a just calling parseText on its content. It uses a syntax of the form ```"key" "value"```, and SLADE's parser expects something like ```"key" = "value";```. So, three options:
1. Parse "manually" with the tokenizer
2. Extend the parser to understand this kind of syntax (it's gonna need parsing modes)
3. Unix and OSX will have to make do with the IWADPATH and IWADDIR environment variables, no Steam autodetection.

**Comment by sirjuddington** [comment]
If the format is as simple as that it wouldn't be too difficult to parse it manually with the tokenizer

**Comment by eevee** [comment]
I watched a new mapper try SLADE for the first time last night and _repeatedly_ make edits in the prop grid, which were then silently lost, because the default is to require you to click the checkmark to commit the changes.

I don't think that's a great default behavior.  At the very least, if you click away from the prop grid when you've made changes, it should ask whether you wanted to save or discard them.

**Comment by Gaerzi** [comment]
I agree with that, having to validate prop grid changes is something I frequently forget to do...

**Comment by eevee** [comment]
There's a pref to auto-apply prop grid changes as soon as you make them.  I'm not sure why that isn't just the default behavior, given that you can always undo changes you didn't want.
